File: java/src/test/java/com/google/oak/transport/GrpcStreamingTransportTest.java
//
// Copyright 2023 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package com.google.oak.transport;

import static org.mockito.AdditionalAnswers.delegatesTo;
import static org.mockito.Mockito.mock;

import com.google.oak.attestation.v1.Endorsements;
import com.google.oak.attestation.v1.Evidence;
import com.google.oak.crypto.v1.EncryptedRequest;
import com.google.oak.crypto.v1.EncryptedResponse;
import com.google.oak.session.v1.EndorsedEvidence;
import com.google.oak.session.v1.GetEndorsedEvidenceResponse;
import com.google.oak.session.v1.InvokeResponse;
import com.google.oak.session.v1.RequestWrapper;
import com.google.oak.session.v1.ResponseWrapper;
import com.google.oak.session.v1.StreamingSessionGrpc;
import com.google.oak.transport.GrpcStreamingTransport;
import com.google.oak.util.Result;
import io.grpc.ManagedChannel;
import io.grpc.inprocess.InProcessChannelBuilder;
import io.grpc.inprocess.InProcessServerBuilder;
import io.grpc.stub.StreamObserver;
import io.grpc.testing.GrpcCleanupRule;
import java.lang.IllegalArgumentException;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class GrpcStreamingTransportTest {
  private static class RequestStreamObserver implements StreamObserver<RequestWrapper> {
    private final StreamObserver<ResponseWrapper> responseObserver;

    RequestStreamObserver(StreamObserver<ResponseWrapper> responseObserver) {
      this.responseObserver = responseObserver;
    }

    @Override
    public void onNext(RequestWrapper request) {
      ResponseWrapper responseWrapper;
      RequestWrapper.RequestCase requestCase = request.getRequestCase();
      switch (requestCase) {
        case GET_ENDORSED_EVIDENCE_REQUEST:
          responseWrapper = ResponseWrapper.newBuilder()
                                .setGetEndorsedEvidenceResponse(
                                    GetEndorsedEvidenceResponse.newBuilder().setEndorsedEvidence(
                                        EndorsedEvidence.newBuilder()
                                            .setEvidence(Evidence.getDefaultInstance())
                                            .setEndorsements(Endorsements.getDefaultInstance())))
                                .build();
          responseObserver.onNext(responseWrapper);
          break;
        case INVOKE_REQUEST:
          responseWrapper = ResponseWrapper.newBuilder()
                                .setInvokeResponse(InvokeResponse.newBuilder().setEncryptedResponse(
                                    EncryptedResponse.getDefaultInstance()))
                                .build();
          responseObserver.onNext(responseWrapper);
          break;
        case REQUEST_NOT_SET:
          responseObserver.onError(new IllegalArgumentException());
          break;
      }
    }

    @Override
    public void onError(Throwable t) {
      responseObserver.onError(t);
    }

    @Override
    public void onCompleted() {
      responseObserver.onCompleted();
    }
  }

  private final StreamingSessionGrpc.StreamingSessionImplBase serviceImpl =
      mock(StreamingSessionGrpc.StreamingSessionImplBase.class,
          delegatesTo(new StreamingSessionGrpc.StreamingSessionImplBase() {
            @Override
            public StreamObserver<RequestWrapper> stream(
                StreamObserver<ResponseWrapper> responseObserver) {
              return new RequestStreamObserver(responseObserver);
            }
          }));

  /**
   * This rule manages automatic graceful shutdown for the registered servers and
   * channels at the end of test.
   */
  @Rule public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();

  private StreamingSessionGrpc.StreamingSessionStub client;

  @Before
  public void setUp() throws Exception {
    // Generate a unique in-process server name.
    String serverName = InProcessServerBuilder.generateName();

    // Create a server, add service, start, and register for automatic graceful
    // shutdown.
    grpcCleanup.register(InProcessServerBuilder.forName(serverName)
                             .directExecutor()
                             .addService(serviceImpl)
                             .build()
                             .start());

    // Create a client channel and register for automatic graceful shutdown.
    ManagedChannel channel =
        grpcCleanup.register(InProcessChannelBuilder.forName(serverName).directExecutor().build());

    // Create a client using the in-process channel;
    client = StreamingSessionGrpc.newStub(channel);
  }

  @Test
  public void testGrpcStreamingTransport() throws Exception {
    GrpcStreamingTransport transport = new GrpcStreamingTransport(client::stream);

    Result<EndorsedEvidence, String> getEndorsedEvidenceResult = transport.getEndorsedEvidence();
    Assert.assertTrue(getEndorsedEvidenceResult.isSuccess());

    Result<EncryptedResponse, String> invokeResult =
        transport.invoke(EncryptedRequest.getDefaultInstance());
    Assert.assertTrue(invokeResult.isSuccess());
    Assert.assertEquals(
        invokeResult.unwrap("missing result"), EncryptedResponse.getDefaultInstance());

    // The following call may throw a general {@code Exception}.
    // The test succeeds if it doesn't throw an exception.
    transport.close();
  }
}

File: cc/transport/grpc_streaming_transport_test.cc
/*
 * Copyright 2023 The Project Oak Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cc/transport/grpc_streaming_transport.h"

#include <chrono>
#include <thread>

#include "absl/log/absl_check.h"
#include "absl/log/log.h"
#include "absl/status/status.h"
#include "gmock/gmock.h"
#include "grpcpp/server_builder.h"
#include "grpcpp/support/time.h"
#include "gtest/gtest.h"
#include "proto/crypto/crypto.pb.h"
#include "proto/session/messages.pb.h"
#include "proto/session/service_streaming.grpc.pb.h"
#include "proto/session/service_streaming.pb.h"

using oak::crypto::v1::EncryptedRequest;
using oak::crypto::v1::EncryptedResponse;
using oak::session::v1::EndorsedEvidence;
using oak::session::v1::RequestWrapper;
using oak::session::v1::ResponseWrapper;
using oak::transport::GrpcStreamingTransport;
using ::testing::_;
using ServerStream =
    ::grpc::ServerReaderWriter<ResponseWrapper, RequestWrapper>;
using ClientStream =
    ::grpc::ClientReaderWriterInterface<RequestWrapper, ResponseWrapper>;

class MockServiceStreaming
    : public ::oak::session::v1::StreamingSession::Service {
 public:
  MOCK_METHOD(grpc::Status, Stream, (grpc::ServerContext*, (ServerStream*)),
              (override));
};
class GrpcStreamingTransportTest : public ::testing::Test {
 protected:
  ::grpc::ClientContext context_;
  std::unique_ptr<grpc::Server> server_;
  std::shared_ptr<grpc::Channel> channel_;
  std::unique_ptr<oak::session::v1::StreamingSession::StubInterface> stub_;
  MockServiceStreaming mock_service_;

  void SetUp() override {
    grpc::ServerBuilder server_builder;
    server_builder.RegisterService(&mock_service_);
    server_ = server_builder.BuildAndStart();
    channel_ = server_->InProcessChannel({});
    stub_ = oak::session::v1::StreamingSession::NewStub(channel_);

    context_.set_deadline(gpr_time_from_seconds(10, GPR_TIMESPAN));
  }

  void TearDown() override {
    server_->Shutdown();
    server_->Wait();
  }
};

TEST_F(GrpcStreamingTransportTest, InvokePropagatesSendError) {
  GrpcStreamingTransport transport(stub_->Stream(&context_));

  EXPECT_CALL(mock_service_, Stream(_, _))
      .WillOnce([](grpc::ServerContext*, ServerStream* stream) {
        return grpc::Status(grpc::StatusCode::FAILED_PRECONDITION,
                            "fake error");
      });

  EncryptedRequest request;
  absl::StatusOr<oak::crypto::v1::EncryptedResponse> response =
      transport.Invoke(request);
  ASSERT_EQ(response.status(),
            absl::Status(absl::StatusCode::kFailedPrecondition,
                         "while writing request: fake error"));
}

TEST_F(GrpcStreamingTransportTest, GetEndorsedEvidencePropagatesSendError) {
  GrpcStreamingTransport transport(stub_->Stream(&context_));

  EXPECT_CALL(mock_service_, Stream(_, _))
      .WillOnce([](grpc::ServerContext*, ServerStream* stream) {
        return grpc::Status(grpc::StatusCode::FAILED_PRECONDITION,
                            "fake error");
      });

  absl::StatusOr<EndorsedEvidence> response = transport.GetEndorsedEvidence();
  ASSERT_EQ(response.status(),
            absl::Status(absl::StatusCode::kFailedPrecondition,
                         "while writing request: fake error"));
}

TEST_F(GrpcStreamingTransportTest, InvokePropagatesWeirdError) {
  GrpcStreamingTransport transport(stub_->Stream(&context_));

  EXPECT_CALL(mock_service_, Stream(_, _))
      .WillOnce([](grpc::ServerContext*, ServerStream* stream) {
        return grpc::Status::OK;
      });

  EncryptedRequest request;
  absl::StatusOr<oak::crypto::v1::EncryptedResponse> response =
      transport.Invoke(request);

  ASSERT_EQ(response.status().code(), absl::StatusCode::kInternal);
  EXPECT_THAT(response.status().message(),
              testing::StartsWith("failed to read request"));
}

TEST_F(GrpcStreamingTransportTest, GetEndorsedEvidencePropagatesWeirdError) {
  ::grpc::ClientContext context;
  GrpcStreamingTransport transport(stub_->Stream(&context));

  EXPECT_CALL(mock_service_, Stream(_, _))
      .WillOnce([](grpc::ServerContext*, ServerStream* stream) {
        return grpc::Status::OK;
      });

  absl::StatusOr<EndorsedEvidence> response = transport.GetEndorsedEvidence();

  ASSERT_EQ(response.status().code(), absl::StatusCode::kInternal);
  EXPECT_THAT(response.status().message(),
              testing::StartsWith("failed to read request"));
}
